src/access/parsers/config.py

```diff
@@ -0,0 +1,420 @@
+
+    The branch should store a value rule of the appropriate type.
+
+    Args:
+        branch (Tree): Tree branch to update.
+        value (Any): New value.
+
+    Raises:
+        TypeError: Raises an exception if the new and old value types do not match.
+    """
+    if not hasattr(branch, "data") or not _value_sanity_check[branch.data](value):
+        raise TypeError("Trying to change value type")
+    transformed_value = _value_inverse_transformers[branch.data](value, branch.children[0])
+    branch.children[0] = branch.children[0].update(value=transformed_value)  # type: ignore
+
+
+class Config(dict):
+    """Class inheriting from dict used to store the contents of parsed configuration files.
+
+    For each entry we keep a reference to the corresponding branch in the parse tree so that we can update it when
+    changing the contents of the dict. This is then done by overriding the __setitem__ and __delitem__ methods.
+
+    The class also adds support for case-insensitive keys by overriding the appropriate dict methods.
+
+    Args:
+        tree (Tree): the parse tree, as returned by Lark.
+        reconstructor (Reconstructor): the Lark reconstructor built from the grammar.
+        case_sensitive_keys (bool): Are keys case-sensitive?
+    """
+
+    _tree: Tree  # The full parse tree
+    _refs: dict  # References to the nodes of the parse tree
+    _reconstructor: Reconstructor  # Lark reconstrutor used for round-trip parsing
+    _case_sensitive_keys: bool  # Are the dict keys case insensitive?
+
+    def __init__(self, tree: Tree, reconstructor: Reconstructor, case_sensitive_keys: bool) -> None:
+        self._tree = tree
+        self._reconstructor = reconstructor
+        self._case_sensitive_keys = case_sensitive_keys
+        interpreter = ConfigToDict(reconstructor, case_sensitive_keys)
+        data, self._refs = interpreter.visit(self._tree)
+        super().__init__(data)
+
+    def __getitem__(self, key: str) -> Any:
+        """Override method to get item from dict.
+
+        This method takes into account if keys are case-sensitive or not."""
+        if self._case_sensitive_keys:
+            return super().__getitem__(key)
+        else:
+            return super().__getitem__(key.upper())
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        """Override method to set item from dict.
+
+        This method takes care of updading the parse tree, so that when writing it back into text it will use the new
+        values. To make sure this works correctly, we check that the type of the new value is consistent with the
+        current type. This method also takes into account if keys are case-sensitive or not."""
+
+        if not self._case_sensitive_keys:
+            key = key.upper()
+
+        # Currently we only support replacing existing values, not adding new ones
+        if key not in self:
+            raise (KeyError("Key already exist: ", key))
```

Comment:
```suggestion
            raise (KeyError(f"Key doesn't exist: {key}"))
```
changes the error message from `('Key already exist: ', 'potato')`
to `KeyError: "Key doesn't exist: potato"`
If adding keys will be supported later on, maybe `NotImplemented` error could be raised instead?

Reply:
Not entirely sure, but I don't think adding keys will be supported. It involves quite a lot of code (creating an entire tree branch for the new item) for limited use-cases.